### ISSUE TITLE
podman machine start cannot find gvproxy when brew is not installed in the standard directory

### ISSUE THREAD

**Issue by carlos-algms** [containers/podman#12161]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

It is not possible to run `podman machine start` when brew is not installed in the default location. 

**Steps to reproduce the issue:**

1. Install `brew` on your user directory, as described in the [official docs](https://docs.brew.sh/Installation#untar-anywhere):
 ```mkdir ~/homebrew && curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C ~/homebrew```

2. Install `podman`
``` 
brew install podman
podman machine init
podman machine start 
```

**Describe the results you received:**

`podman machine start` returns: 
```
Error: unable to start host networking: 
"could not find \"gvproxy\" in one of 
[
  /usr/local/opt/podman/libexec 
  /opt/homebrew/bin /opt/homebrew/opt/podman/libexec 
  /usr/local/bin /usr/local/libexec/podman 
  /usr/local/lib/podman 
  /usr/libexec/podman 
  /usr/lib/podman
]"
```

**Describe the results you expected:**
Podman should be able to find its install location and pick up the `gvproxy` bin inside its `libexec` folder. 

I can confirm that `gvproxy` IS inside `~/homebrew/Cellar/podman/3.4.1/libexec/` folder.

**Output of `podman version`:**

```
Cannot connect to Podman ....
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/master/troubleshooting.md)**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**
MacOS Big Sur 11.3.1

**Comment by Luap99** [comment]
see https://github.com/containers/podman/issues/11960
I think this should be reported against homebrew, there is no way that podman can know about custom install locations.
@ashley-cui PTAL

**Comment by carlos-algms** [comment]
I reported it here because I thought it was podman installing and executing `gvproxy` as a dependency under the `libexec` folder.

**Comment by ashley-cui** [comment]
gvproxy is installed by homebrew as a dependency in the podman libexec folder. homebrew does not recommend installing in another directory, and in the docs there is a warning that things may break when homebrew is installed in another place.

**Comment by carlos-algms** [comment]
I know, but `gvproxy` is the only missing part, if I link it to one of the paths podman listed that it is looking for it, everything works as expected. 

Can it then be a new feature?
Where podman just reads its `realpath` and look for `gvproxy` in one more folder? 
Its own **libexec** folder? 

It would be great for people that do not have `root` access, as it is blocked by the Enterprise restrictions, so installing in other places is the only alternative. 

Thanks.

**Comment by afbjorklund** [comment]
Normally you wouldn't link directly to the Cellar, but to the prefix: `~/homebrew/libexec`

But the default configuration still has to be patched, since it only has hardcoded locations.

**Comment by ashley-cui** [comment]
@afbjorklund for most places, it would link to `$HOMEBREW_PREFIX/bin` or `$HOMEBREW_PREFIX/lib`, but libexec is special in Homebrew, it doesn't get linked to the homebrew prefix

**Comment by afbjorklund** [comment]
> but libexec is special in Homebrew, it doesn't get linked to the homebrew prefix

right, because users are supposed to only call the stuff that is available in bin

**Comment by afbjorklund** [comment]
Unfortunately, it looks like the formula is broken.

```ruby
    resource("gvproxy").stage do
      system "make", "gvproxy"
      libexec.install "bin/gvproxy"
    end
```

It installs /usr/local/libexec/gvproxy et. al. :-(

It was supposed to be in a subdirectory ?

**Comment by ashley-cui** [comment]
`#{prefix}` is `/usr/local/Cellar/foo/0.1/`. `$HOMEBREW_PREFIX` is different, it's `/usr/local`. 
every formula in cellar is also symlinked into `$HOMEBREW_PREFIX/opt/foo`.

**Comment by afbjorklund** [comment]
Right, here they translate to:

HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew

prefix: /home/linuxbrew/.linuxbrew/Cellar/podman/3.4.1

**Comment by rhatdan** [comment]
Where do we stand on this issue?  Is it still a problem?
@ashley-cui @afbjorklund @carlos-algms WDYT?

**Comment by ashley-cui** [comment]
I don't think this is an issue that should be fixed on our end, its more of a homebrew install issue, and homebrew specifically states that some packages may break if not installed in their default location

**Comment by afbjorklund** [comment]
They forgot to patch the default config, with the current path to use.

**Comment by afbjorklund** [comment]
Back in the day we used `./configure`. Not sure what the new variant would be?

**Comment by carlos-algms** [comment]
I would say it is good for the community in general to have this fixed. 
As it was pointed out, the paths are hardcoded and It would be great to have a dynamic check on the installation folder before giving up looking for executables.

I raised this ticket because if your computer is blocked by the company work, and you don't have admin/sudo rights, installing on a custom directory is the only option.

**Comment by afbjorklund** [comment]
You can provide your custom directory in configuration, if Homebrew doesn't want to fix their packaging

**Comment by afbjorklund** [comment]
I patched out all the hardcoded paths, but it seems like they were vendored from containers-common ?

**Comment by afbjorklund** [comment]
Homebrew prefers to have their packages unpatched, so left all the unused paths in there.

/usr/local/opt/podman/libexec
/opt/homebrew/bin
/opt/homebrew/opt/podman/libexec
/usr/local/bin
**/usr/local/libexec/podman**
/usr/local/lib/podman
/usr/libexec/podman
/usr/lib/podman

**Comment by carlocab** [comment]
> I don't think this is an issue that should be fixed on our end, its more of a homebrew install issue

It seems weird to call this a Homebrew install issue when this is also causing problems for user who aren't even using Homebrew. See #11960 (also referenced above).

**Comment by afbjorklund** [comment]
Technically it is a problem for all "weird" installations that don't use the "usual" `/usr` (or `/usr/local`).

But there really isn't any good substitute for `./configure --prefix=$PREFIX` either, so needs patching...

**Comment by bowbahdoe** [comment]
This is an issue for nix installs

**Comment by maximsmol** [comment]
as a workaround, add this to `~/.config/containers/containers.conf`:
```toml
[engine]
helper_binaries_dir = [
  "{PATH_TO_GPROXY_BIN}",
  "/usr/local/libexec/podman",
  "/usr/local/lib/podman",
  "/usr/libexec/podman",
  "/usr/lib/podman",
]
```

on my Mac system using Nix the path is `/nix/store/k92v259vn6bv0shsifxhjvavbj436906-gvproxy-0.3.0/bin`
you could probably generate this via home-manager but there are other settings that get in the way

I'll work on a PR for nixpkgs soon

**Comment by afbjorklund** [comment]
It could be a problem to set one "helper" dir, if you spread the podman helpers into different packages.

On some system they have relocated them into the podman libexec, while others use a global libexec.

**Comment by carlocab** [comment]
@ashley-cui I think at this stage it's fairly clear that this issue isn't specific to Homebrew.

Perhaps you'd reconsider implementing a solution in this project? It seems really inefficient for your various downstream distributors to have to carry around duplicate pieces of code to work around this issue.

**Comment by ashley-cui** [comment]
@carlocab I'll bring it up with the team and see what the consensus is. I think we've gone back and forth a lot on where this gvproxy binary should live, but I can bring it up again.

**Comment by carlocab** [comment]
I don't think it's really a matter of where it should live: I think the above thread shows that that's something difficult to control.

To me, it's a matter of providing a mechanism for telling `podman` where `gvproxy` will live and documenting this. Whether this mechanism is only invoked at build-time or is also available at run-time is up to you.

**Comment by bowbahdoe** [comment]
@carlocab I will add that there is no way i have found to resolve this when installing podman via nix

**Comment by carlocab** [comment]
@bowbahdoe something that works for us currently is to replace these lines:

https://github.com/containers/podman/blob/0cfb5d7e75507783e463e0df645e873e5a77bbfa/vendor/github.com/containers/common/pkg/config/config_darwin.go#L26

https://github.com/containers/podman/blob/0cfb5d7e75507783e463e0df645e873e5a77bbfa/vendor/github.com/containers/common/pkg/config/config_linux.go#L40

with the absolute path to directory we expect to contain `gvproxy`. You can see this [here](https://github.com/Homebrew/homebrew-core/blob/c5a89f1a8c111a1309deecb46f97b5b02fe38ebe/Formula/podman.rb#L42-L44). That may or may not be something you can also do in nix, depending on when and on what the hashes used to determine the install path is computed. (I don't use nix, so I don't actually know.)

**Comment by maximsmol** [comment]
@bowbahdoe I'm guessing you tried adding my example path? If your nixpkgs is at all different you will get a different derivation and output path. You can get yours by running `nix repl '<nixpkgs>'` and taking the value of `pkgs.gvproxy.outPath`. If that's not it, would be great to see what error you are getting so we can figure out whether this is really blocking in some cases on nix

Example:
```
$ nix repl '<nixpkgs>'
Welcome to Nix 2.6.0. Type :? for help.

Loading '<nixpkgs>'...
Added 15937 variables.

nix-repl> pkgs.gvproxy.outPath
"/nix/store/k92v259vn6bv0shsifxhjvavbj436906-gvproxy-0.3.0"
```

**Comment by ashley-cui** [comment]
Alright, we'll look into a solution where a location can be set during compile-time using our makefile.

**Comment by afbjorklund** [comment]
I don't understand how Nix can work, without some PATH directory to put the symlinks in (like brew) ?

Surely you don't add the installation directory for each and every little package, to the global variable ?

**Comment by afbjorklund** [comment]
@ashley-cui : afaik, you already set `config._installPrefix` from the Makefile ?

```Makefile
PREFIX ?= /usr/local
BINDIR ?= ${PREFIX}/bin
LIBEXECDIR ?= ${PREFIX}/libexec
LIBEXECPODMAN ?= ${LIBEXECDIR}/podman
MANDIR ?= ${PREFIX}/share/man
SHAREDIR_CONTAINERS ?= ${PREFIX}/share/containers
ETCDIR ?= ${PREFIX}/etc
```
```Makefile
	-X $(LIBPOD)/config._installPrefix=$(PREFIX) \
```

So need to do something similar for the helpers (`$(LIBEXECPODMAN)`), perhaps.

**Comment by ashley-cui** [comment]
Closed via: #13372

Example usage: `make podman-remote HELPER_BINARIES_DIR=/my/location/prefix`

**Comment by deadpyxel** [comment]
I'll check that. It seems `homebrew` is installing the latest version, but maybe it is something with the package as you say.
